Coursera IBM Data Scicence

Course 4 Python for Data Science, AI & Development

Module 1 Python Basics

Hands-On Lab: Your First Program, Types, Expressions, and Variables

<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="Skills Network Logo">
    </a>
</p>


# Writing Your First Python Code


Estimated time needed: **30** minutes
    

## Objectives

After completing this lab you will be able to:

* Write basic code in Python
* Work with various types of data in Python
* Convert the data from one type to another
* Use expressions and variables to perform operations


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>
            <a href="#Say-'Hello'-to-the-world-in-Python">Say 'Hello' to the world in Python</a>
            <ul>
                <li><a href="#What-version-of-Python-are-we-using?">What version of Python are we using?</a></li>
                <li><a href="#Writing-comments-in-Python">Writing comments in Python</a></li>
                <li><a href="#Errors-in-Python">Errors in Python</a></li>
                <li><a href="#Does-Python-know-about-your-error-before-it-runs-your-code?">Does Python know about your error before it runs your code?</a></li>
                <li><a href="#Exercise:-Your-First-Program">Exercise: Your First Program</a></li>
            </ul>
        </li>
        <li>
            <a href="#Types-of-objects-in-Python">Types of objects in Python</a>
            <ul>
                <li><a href="#Integers">Integers</a></li>
                <li><a href="#Floats">Floats</a></li>
                <li><a href="#Converting-from-one-object-type-to-a-different-object-type">Converting from one object type to a different object type</a></li>
                <li><a href="#Boolean-data-type">Boolean data type</a></li>
                <li><a href="#Exercise:-Types">Exercise: Types</a></li>
            </ul>
        </li>
        <li>
            <a href="#Expressions-and-Variables">Expressions and Variables</a>
            <ul>
                <li><a href="#Expressions">Expressions</a></li>
                <li><a href="#Exercise:-Expressions">Exercise: Expressions</a></li>
                <li><a href="#Variables">Variables</a></li>
                <li><a href="#Exercise:-Expression-and-Variables-in-Python">Exercise: Expression and Variables in Python</a></li>
            </ul>
        </li>
    </ul>
</div>

<hr>


## Say 'Hello' to the world in Python


When learning a new programming language, it is customary to start with an "hello world" example. As simple as it is, this one line of code will ensure that we know how to print a string in output and how to execute code within cells in a notebook.


<hr/>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
[Tip]: To execute the Python code in the code cell below, click on the cell to select it and press <kbd>Shift</kbd> + <kbd>Enter</kbd>.
</div>
<hr/>


In [1]:
# Try your first Python output

print('Hello, Python!')

Hello, Python!


After executing the cell above, you should see that Python prints <code>Hello, Python!</code>. Congratulations on running your first Python code!


<hr/>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
    [Tip:] <code>print()</code> is a function. You passed the string <code>'Hello, Python!'</code> as an argument to instruct Python on what to print.
</div>
<hr/>


### What version of Python are we using?


<p>
    There are two popular versions of the Python programming language in use today: Python 2 and Python 3. The Python community has decided to move on from Python 2 to Python 3, and many popular libraries have announced that they will no longer support Python 2.
</p>
<p>
    Since Python 3 is the future, in this course we will be using it exclusively. How do we know that our notebook is executed by a Python 3 runtime? We can look in the top-right hand corner of this notebook and see "Python 3".
</p>
<p>
    We can also ask Python directly and obtain a detailed answer. Try executing the following code:
</p>


In [2]:
# Check the Python Version

import sys
print(sys.version)

3.10.12 (main, Nov  6 2024, 20:22:13) [GCC 11.4.0]


<hr/>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
    [Tip:] <code>sys</code> is a built-in module that contains many system-specific parameters and functions, including the Python version in use. Before using it, we must explictly <code>import</code> it.
</div>
<hr/>


### Writing comments in Python


<p>
    In addition to writing code, note that it's always a good idea to add comments to your code. It will help others understand what you were trying to accomplish (the reason why you wrote a given snippet of code). Not only does this help <strong>other people</strong> understand your code, it can also serve as a reminder <strong>to you</strong> when you come back to it weeks or months later.</p>

<p>
    To write comments in Python, use the number symbol <code>#</code> before writing your comment. When you run your code, Python will ignore everything past the <code>#</code> on a given line.
</p>


In [3]:
# Practice on writing comments

print('Hello, Python!') # This line prints a string
# print('Hi')

Hello, Python!


<p>
    After executing the cell above, you should notice that <code>This line prints a string</code> did not appear in the output, because it was a comment (and thus ignored by Python).
</p>
<p>
    The second line was also not executed because <code>print('Hi')</code> was preceded by the number sign (<code>#</code>) as well! Since this isn't an explanatory comment from the programmer, but an actual line of code, we might say that the programmer <em>commented out</em> that second line of code.
</p>


<h3 id="errors">Errors in Python</h3>


<p>Everyone makes mistakes. For many types of mistakes, Python will tell you that you have made a mistake by giving you an error message. It is important to read error messages carefully to really understand where you made a mistake and how you may go about correcting it.</p>
<p>For example, if you spell <code>print</code> as <code>frint</code>, Python will display an error message. Give it a try:</p>


In [4]:
# Print string as error message

frint("Hello, Python!")

NameError: name 'frint' is not defined

<p>The error message tells you:
<ol>
    <li>where the error occurred (more useful in large notebook cells or scripts), and</li>
    <li>what kind of error it was (NameError)</li>
</ol>
<p>Here, Python attempted to run the function <code>frint</code>, but could not determine what <code>frint</code> is since it's not a built-in function and it has not been previously defined by us either.</p>


<p>
    You'll notice that if we make a different type of mistake, by forgetting to close the string, we'll obtain a different error (i.e., a <code>SyntaxError</code>). Try it below:
</p>


In [35]:
# Try to see built-in error message

print("Hello, Python!)

SyntaxError: unterminated string literal (detected at line 3) (<ipython-input-35-f0b5a635e1a2>, line 3)

### Does Python know about your error before it runs your code?


Python is what is called an <em>interpreted language</em>. Compiled languages examine your entire program at compile time, and are able to warn you about a whole class of errors prior to execution. In contrast, Python interprets your script line by line as it executes it. Python will stop executing the entire program when it encounters an error (unless the error is expected and handled by the programmer, a more advanced subject that we'll cover later on in this course).


Try to run the code in the cell below and see what happens:


In [36]:
# Print string and error to see the running order

print("This will be printed")
frint("This will cause an error")
print("This will NOT be printed")

This will be printed


NameError: name 'frint' is not defined

### Exercise: Your First Program


<p>Generations of programmers have started their coding careers by simply printing "Hello, world!". You will be following in their footsteps.</p>
<p>In the code cell below, use the <code>print()</code> function to print out the phrase: <code>Hello, world!</code></p>


In [53]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
print("Hello, world!")

Hello, world!


<details><summary>Click here for the solution</summary>

```python
print("Hello, world!")

```

</details>


<p>Now, let's enhance your code with a comment. In the code cell below, print out the phrase: <code>Hello, world!</code> and comment it with the phrase <code>Print the traditional hello world</code> all in one line of code.</p>


In [54]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
print("Hello, world!") # Print the traditional hello world

Hello, world!


<details><summary>Click here for the solution</summary>

```python
print("Hello, world!") # Print the traditional hello world

```

</details>


<hr>


## Types of objects in Python


<p>Python is an object-oriented language. There are many different types of objects in Python. Let's start with the most common object types: <i>strings</i>, <i>integers</i> and <i>floats</i>. Anytime you write words (text) in Python, you're using <i>character strings</i> (strings for short). The most common numbers, on the other hand, are <i>integers</i> (e.g. -1, 0, 100) and <i>floats</i>, which represent real numbers (e.g. 3.14, -42.0).</p>


<a align="center">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%201/images/TypesObjects.png" width="600">
</a>


<p>The following code cells contain some examples.</p>


In [55]:
# Integer

11

11

In [56]:
# Float

2.14

2.14

In [57]:
# String

"Hello, Python 101!"

'Hello, Python 101!'

<p>You can get Python to tell you the type of an expression by using the built-in <code>type()</code> function. You'll notice that Python refers to integers as <code>int</code>, floats as <code>float</code>, and character strings as <code>str</code>.</p>


In [58]:
# Type of 12

type(12)

int

In [59]:
# Type of 2.14

type(2.14)

float

In [60]:
# Type of "Hello, Python 101!"

type("Hello, Python 101!")

str

<p>In the code cell below, use the <code>type()</code> function to check the object type of <code>12.0</code>.


In [61]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
type(12.0)

float

<details><summary>Click here for the solution</summary>

```python
type(12.0)

```

</details>


### Integers


<p>Here are some examples of integers. Integers can be negative or positive numbers:</p>


<a align="center">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%201/images/TypesInt.png" width="600">
</a>


<p>We can verify this is the case by using, you guessed it, the <code>type()</code> function:


In [62]:
# Print the type of -1

type(-1)

int

In [63]:
# Print the type of 4

type(4)

int

In [64]:
# Print the type of 0

type(0)

int

### Floats


<p>Floats represent real numbers; they are a superset of integer numbers but also include "numbers with decimals". There are some limitations when it comes to machines representing real numbers, but floating point numbers are a good representation in most cases. You can learn more about the specifics of floats for your runtime environment, by checking the value of <code>sys.float_info</code>. This will also tell you what's the largest and smallest number that can be represented with them.</p>

<p>Once again, can test some examples with the <code>type()</code> function:


In [65]:
# Print the type of 1.0

type(1.0) # Notice that 1 is an int, and 1.0 is a float

float

In [66]:
# Print the type of 0.5

type(0.5)

float

In [67]:
# Print the type of 0.56

type(0.56)

float

In [68]:
# System settings about float type

sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

### Converting from one object type to a different object type


<p>You can change the type of the object in Python; this is called typecasting. For example, you can convert an <i>integer</i> into a <i>float</i> (e.g. 2 to 2.0).</p>
<p>Let's try it:</p>


In [69]:
# Verify that this is an integer

type(2)

int

#### Converting integers to floats
<p>Let's cast integer 2 to float:</p>


In [70]:
# Convert 2 to a float

float(2)

2.0

In [71]:
# Convert integer 2 to a float and check its type

type(float(2))

float

<p>When we convert an integer into a float, we don't really change the value (i.e., the significand) of the number. However, if we cast a float into an integer, we could potentially lose some information. For example, if we cast the float 1.1 to integer we will get 1 and lose the decimal information (i.e., 0.1):</p>


In [72]:
# Casting 1.1 to integer will result in loss of information

int(1.1)

1

<h4>Converting from strings to integers or floats</h4>


<p>Sometimes, we can have a string that contains a number within it. If this is the case, we can cast that string that represents a number into an integer using <code>int()</code>:</p>


In [73]:
# Convert a string into an integer

int('1')

1

<p>But if you try to do so with a string that is not a perfect match for a number, you'll get an error. Try the following:</p>


In [74]:
# Convert a string into an integer with error

int('1 or 2 people')

ValueError: invalid literal for int() with base 10: '1 or 2 people'

<p>You can also convert strings containing floating point numbers into <i>float</i> objects:</p>


In [75]:
# Convert the string "1.2" into a float

float('1.2')

1.2

<hr/>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
    [Tip:] Note that strings can be represented with single quotes (<code>'1.2'</code>) or double quotes (<code>"1.2"</code>), but you can't mix both (e.g., <code>"1.2'</code>).
</div>
<hr/>


<h4>Converting numbers to strings</h4>


<p>If we can convert strings to numbers, it is only natural to assume that we can convert numbers to strings, right?</p>


In [76]:
# Convert an integer to a string

str(1)


'1'

<p>And there is no reason why we shouldn't be able to make floats into strings as well:</p>


In [77]:
# Convert a float to a string

str(1.2)

'1.2'

### Boolean data type


<p><i>Boolean</i> is another important type in Python. An object of type <i>Boolean</i> can take on one of two values: <code>True</code> or <code>False</code>:</p>


In [78]:
# Value true

True

True

<p>Notice that the value <code>True</code> has an uppercase "T". The same is true for <code>False</code> (i.e. you must use the uppercase "F").</p>


In [79]:
# Value false

False

False

<p>When you ask Python to display the type of a boolean object it will show <code>bool</code> which stands for <i>boolean</i>:</p>


In [80]:
# Type of True

type(True)

bool

In [81]:
# Type of False

type(False)

bool

<p>We can cast boolean objects to other data types. If we cast a boolean with a value of <code>True</code> to an integer or float we will get a one. If we cast a boolean with a value of <code>False</code> to an integer or float we will get a zero. Similarly, if we cast a 1 to a Boolean, you get a <code>True</code>. And if we cast a 0 to a Boolean we will get a <code>False</code>. Let's give it a try:</p>


In [82]:
# Convert True to int

int(True)

1

In [83]:
# Convert 1 to boolean

bool(1)

True

In [84]:
# Convert 0 to boolean

bool(0)

False

In [85]:
# Convert True to float

float(True)

1.0

### Exercise: Types


<p>What is the data type of the result of: <code>6 / 2</code>?</p>


In [86]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
type(6/2) # float

float

<details><summary>Click here for the solution</summary>

```python
type(6/2) # float

```

</details>


<p>What is the type of the result of: <code>6 // 2</code>? (Note the double slash <code>//</code>.)</p>


In [87]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
type(6//2) # int, as the double slashes stand for integer division

int

<details><summary>Click here for the solution</summary>

```python
type(6//2) # int, as the double slashes stand for integer division

```

</details>


<hr>


## Expressions and Variables


### Expressions


<p>Expressions in Python can include operations among compatible types (e.g., integers and floats). For example, basic arithmetic operations like adding multiple numbers:</p>


In [88]:
# Addition operation expression

43 + 60 + 16 + 41

160

<p>We can perform subtraction operations using the minus operator. In this case the result is a negative number:</p>


In [89]:
# Subtraction operation expression

50 - 60

-10

<p>We can do multiplication using an asterisk:</p>


In [90]:
# Multiplication operation expression

5 * 5

25

<p>We can also perform division with the forward slash:


In [91]:
# Division operation expression

25 / 5

5.0

In [92]:
# Division operation expression

25 / 6

4.166666666666667

<p>As seen in the quiz above, we can use the double slash for integer division, where the result is rounded down to the nearest integer:


In [93]:
# Integer division operation expression

25 // 5

5

In [94]:
# Integer division operation expression

25 // 6

4

### Exercise: Expressions


<p>Let's write an expression that calculates how many hours there are in 160 minutes:


In [108]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
160//60

2

<details><summary>Click here for the solution</summary>

```python
160/60

# Or

160//60

```

</details>


<p>Python follows well accepted mathematical conventions when evaluating mathematical expressions. In the following example, Python adds 30 to the result of the multiplication (i.e., 120).


In [96]:
# Mathematical expression

30 + 2 * 60

150

<p>And just like mathematics, expressions enclosed in parentheses have priority. So the following multiplies 32 by 60.


In [97]:
# Mathematical expression

(30 + 2) * 60

1920

### Variables


<p>Just like with most programming languages, we can store values in <i>variables</i>, so we can use them later on. For example:</p>


In [98]:
# Store value into variable

x = 43 + 60 + 16 + 41

<p>To see the value of <code>x</code> in a Notebook, we can simply place it on the last line of a cell:</p>


In [99]:
# Print out the value in variable

x

160

<p>We can also perform operations on <code>x</code> and save the result to a new variable:</p>


In [100]:
# Use another variable to store the result of the operation between variable and value

y = x / 60
y

2.6666666666666665

<p>If we save a value to an existing variable, the new value will overwrite the previous value:</p>


In [101]:
# Overwrite variable with new value

x = x / 60
x

2.6666666666666665

<p>It's a good practice to use meaningful variable names, so you and others can read the code and understand it more easily:</p>


In [102]:
# Name the variables meaningfully

total_min = 43 + 42 + 57 # Total length of albums in minutes
total_min

142

In [103]:
# Name the variables meaningfully

total_hours = total_min / 60 # Total length of albums in hours
total_hours

2.3666666666666667

<p>In the cells above we added the length of three albums in minutes and stored it in <code>total_min</code>. We then divided it by 60 to calculate total length <code>total_hours</code> in hours. You can also do it all at once in a single expression, as long as you use parenthesis to add the albums length before you divide, as shown below.</p>


In [104]:
# Complicate expression

total_hours = (43 + 42 + 57) / 60  # Total hours in a single expression
total_hours

2.3666666666666667

<p>If you'd rather have total hours as an integer, you can of course replace the floating point division with integer division (i.e., <code>//</code>).</p>


### Exercise: Expression and Variables in Python


<p>What is the value of <code>x</code> where <code>x = 3 + 2 * 2</code></p>


In [105]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
x = 3 + 2 * 2
x

7

<details><summary>Click here for the solution</summary>

```python
7

```

</details>


<p>What is the value of <code>y</code> where <code>y = (3 + 2) * 2</code>?</p>


In [106]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
y = (3 + 2) * 2
y

10

<details><summary>Click here for the solution</summary>

```python
10

```

</details>


<p>What is the value of <code>z</code> where <code>z = x + y</code>?</p>


In [107]:
# Write your code below. Don't forget to press Shift+Enter to execute the cell
z = x + y
z

17

<details><summary>Click here for the solution</summary>

```python
17

```

</details>


<hr>
<h2>The last exercise!</h2>
<p>Congratulations, you have completed your first lesson and hands-on lab in Python.
<hr>


## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/" target="_blank">Joseph Santarcangelo</a>


## Other contributors

<a href="www.linkedin.com/in/jiahui-mavis-zhou-a4537814a">Mavis Zhou</a>


## Change Log


|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |
|---|---|---|---|
| 2023-10-30 | 2.2 | Abhishek Gagneja| Updated instructions |
| 2022-01-10  | 2.1  | Malika  | Removed the readme for GitShare|
| 2020-08-26  | 2.0  | Lavanya | Moved lab to course repo in GitLab |



## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
